File: backend/app/api/problems.py
from fastapi import APIRouter, HTTPException

from app.services.problem_service import get_all_problems, get_problem_solution

router = APIRouter()

# Fields safe to expose in list view (no solution_query)
LIST_FIELDS = [
    "id", "slug", "title", "difficulty", "category", "dataset",
    "description", "schema_hint", "concept_tags", "hints",
    "explanation", "approach", "common_mistakes",
]


@router.get("")
@router.get("/")
async def list_problems(
    dataset: str | None = None,
    difficulty: str | None = None,
    category: str | None = None,
):
    problems = get_all_problems(dataset=dataset, difficulty=difficulty, category=category)
    total = len(problems)
    safe = [{k: p[k] for k in LIST_FIELDS if k in p} for p in problems]
    return {"problems": safe, "total": total}


@router.get("/{slug}")
async def get_problem(slug: str):
    problem = get_problem_solution(slug)
    if not problem:
        raise HTTPException(status_code=404, detail="Problem not found")
    # Never expose solution_query to frontend
    safe = {k: v for k, v in problem.items() if k != "solution_query"}
    return safe

File: frontend/src/lib/problems.ts
/**
 * Static problem data — embedded at build time so the practice arena,
 * interview simulator, and profile pages work without a live backend.
 * Only query execution requires the backend.
 */

import problemsData from "@/data/problems.json";

export interface ProblemListItem {
  id: string;
  slug: string;
  title: string;
  difficulty: string;
  category: string;
  dataset: string;
  concept_tags: string[];
}

export interface ProblemDetail extends ProblemListItem {
  description: string;
  schema_hint: string[];
  hints: string[];
  explanation: string;
  approach: string | string[];
  common_mistakes: string[];
}

const problems = problemsData.problems as ProblemDetail[];

/** Get all problems (optionally filtered). */
export function getAllProblems(filters?: {
  dataset?: string;
  difficulty?: string;
  category?: string;
}): ProblemDetail[] {
  if (!filters) return problems;
  return problems.filter((p) => {
    if (filters.dataset && p.dataset !== filters.dataset) return false;
    if (filters.difficulty && p.difficulty !== filters.difficulty) return false;
    if (filters.category && p.category !== filters.category) return false;
    return true;
  });
}

/** Get a single problem by slug. */
export function getProblemBySlug(slug: string): ProblemDetail | undefined {
  return problems.find((p) => p.slug === slug);
}

/** Get total count. */
export function getProblemCount(): number {
  return problems.length;
}
